### DISCUSSION TITLE
[K8S] livenessProbe and readinessProbe for celery beat and workers

### DISCUSSION THREAD

**Discussion by JorisAndrade** [celery/celery#10166] [Ideas]
Hi,

I'm using Kubernetes to deploy my python application, Kubernetes provide a livenessProbe and readinessProbe [see here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/). 

How can I do to check if my celery beat or celery worker is alive and in correct state ?
The PID is not a solution because it cannot be used to catch a deadlock for example.

Thanks in advance for your help,

Best regards,

**Comment by thedrow** [comment]
Celery has a [monitoring API](http://docs.celeryproject.org/en/latest/userguide/monitoring.html#worker-events) you can use.
A pod should be considered live if the Celery worker sends a [heartbeat](http://docs.celeryproject.org/en/latest/userguide/monitoring.html#worker-heartbeat).
A pod should be considered ready if the worker has sent the [worker-online](http://docs.celeryproject.org/en/latest/userguide/monitoring.html#worker-online) event.

If you have specific problems or feature requests, please open a separate issue.

**Comment by ScottEAdams** [comment]
Would this work?
```
readinessProbe:
          exec:
            command:
            - "/bin/sh"
            - "-c"
            - "celery -A path.to.app status | grep -o ': OK'"
          initialDelaySeconds: 30
          periodSeconds: 10
```

**Comment by thedrow** [comment]
@7wonders You'd need to extract the celery node name first. This readinessProbe will fail if any celery instance fails which is not what you want.

**Comment by ScottEAdams** [comment]
@thedrow Hmm, I think its actually that it will succeed even if the actual node has failed but another one is ok which is also not a great outcome.

**Comment by redbaron** [comment]
Looks like 

`/bin/sh -c 'exec celery -A path.to.app inspect ping -d celery@$HOSTNAME'` is good enough for readiness check and verifies just one node.

**Comment by desaintmartin** [comment]
Beware that in some apps, running this command can take a few seconds using full CPU AND kubernetes defaults are to run it every 10 seconds.

It is thus much safer to have a high periodSeconds (ours is set to 300).

**Comment by rana-ahmed** [comment]
@redbaron did that command work for you? If it works then what are the settings for liveness and readiness prob?

**Comment by msta** [comment]
For some reason, this readiness probe is nowhere near satisfactory for us. The inspect responds non-deterministically with no load on our cluster. We run the format like this:

celery inspect ping -b "redis://archii-redis-master:6379" -d celery@archii-task-crawl-integration-7d96d86b9d-jwtq7

And with normal ping times (10 seconds), our cluster is completely killed by the CPU celery requires.

**Comment by joekohlsdorf** [comment]
~I use this for liveness with a 30s interval: `sh -c celery -A path.to.app status | grep "${HOSTNAME}:.*OK"`~
I use this for liveness with a 30s interval: `sh -c celery -A path.to.app inspect ping --destination celery@${HOSTNAME}`
Doesn't seem to cause any extra load, I run a fleet of well over 100 workers.

Readiness probes aren't necessary, Celery is never used in a service. I just set `minReadySeconds: 10` which is good enough for delaying worker startup in rolling Deployments, but it obviously depends on the startup time of Celery for your project so examine logs and set accordingly.

**Comment by WillPlatnick** [comment]
Readiness probes are still useful even if they're not used in a service. Specifically, when you do a deployment of workers and  want to make sure your deployment was successful, you usually use ```kubectl rollout status deployment```. Without readinessprobes, we've deployed bad code that didn't start celery and didn't know it.

**Comment by yardensachs** [comment]
My solution was:
```yaml
readinessProbe:
  exec:
    command:
      [
        "/usr/local/bin/python",
        "-c",
        "\"import os;from celery.task.control import inspect;from <APP> import celery_app;exit(0 if os.environ['HOSTNAME'] in ','.join(inspect(app=celery_app).stats().keys()) else 1)\""
      ]
```
Others seem to not work 🤷‍♂️

**Comment by Strangerxxx** [comment]
Thanks @yardensachs!
Spend many time for debugging what's wrong with other solutions, but no way
Seems like `celery inspect ping` command do not return exit(0) or something in that way

**Comment by mariusburfey** [comment]
`celery inspect ping` does work, but you need `bash` to replace the environment variable like this:
```
        livenessProbe:
          exec:
            # bash is needed to replace the environment variable
            command: [
              "bash",
              "-c",
              "celery inspect ping -A apps -d celery@$HOSTNAME"
            ]
          initialDelaySeconds: 30  # startup takes some time
          periodSeconds: 60  # default is quite often and celery uses a lot cpu/ram then.
          timeoutSeconds: 10  # default is too low
```

**Comment by bwanglzu** [comment]
good to know

**Comment by WillPlatnick** [comment]
We ended up ripping celery inspect ping out from our liveness probes because we found that under heavier load, the ping would just hang for minutes at a time even though the jobs were processing fine and there was no backlog. I have a feeling it had something to do with using eventlet, but we're continuing to look into it.

**Comment by thedrow** [comment]
@WillPlatnick That won't happen with 5.0 because Celery will be async so there will be reserved capacity for control coroutines.

**Comment by shshe** [comment]
I'm having trouble with `inspect ping` spawning defunct / zombie processes:

```
root      2296  0.0  0.0      0     0 ?        Z    16:04   0:00 [python] <defunct>
root      2323  0.0  0.0      0     0 ?        Z    16:05   0:00 [python] <defunct>
...
```

Anyone else encountering this? There isn't a `--pool` argument for forcing a single process execution.

**Comment by mcyprian** [comment]
Can I ask what are you using instead of `celery inspect ping` @WillPlatnick? We've encountered a similar issue with the probe failing under heavy load.

**Comment by WillPlatnick** [comment]
@mcyprian We got rid of the liveness probe. My gut is telling me it has something to do with eventlet, but we haven't made it a priority to figure it out.

**Comment by HeathLee** [comment]
we meet the same CPU problem with Redis broker

**Comment by bartoszhernas** [comment]
Has anyone found a solution? 
We were also experimenting with scheduling "debug_task" on the queue which name we based on the container name. The problem is that we have tons of old queues in RabbitMQ now

**Comment by capital-G** [comment]
Please be aware that using

```
sh -c celery -A path.to.app status | grep "${HOSTNAME}:.*OK"
```

as suggested on https://github.com/celery/celery/issues/4079#issuecomment-437415370 will lead to massive error reports on rabbitmq, see https://github.com/celery/celery/issues/4355#issuecomment-578786369

**Comment by rishi1111** [comment]
I think i have found a way to reduce the CPU usage of inspect ping.

celery -b amqp://user:pass@rabbitmq:5672/vhost inspect ping

Not loading the celery configs using -A path.to.celery certainly helped with the cpu usage,
could somebody verify.

**Comment by desaintmartin** [comment]
> I think i have found a way to reduce the CPU usage of inspect ping.
> 
> celery -b amqp://user:pass@rabbitmq:5672/vhost inspect ping
> 
> Not loading the celery configs using -A path.to.celery certainly helped with the cpu usage,
> could somebody verify.

Nice! It's much better than with the app loaded.
But we still get the huge overhead of a python process starting + celery import. I still would recommend a high period.

**Comment by kweikwei** [comment]
Hi,
celery inspect ping -A app.tasks -d celery@$HOSTNAME gives me "Error: Broadcast not supported by transport 'sqs'" . 
I'm using SQS as the broker, so this means 'inspect' / 'status' command wont work with SQS?

**Comment by innovia** [comment]
We have found that at scale, all the remote control features are causing the Redis instance to spike up in CPU due to set commands on `kombu.pidbox` key, so we can't use ping, or status or inspect as they are all using remote control and trying to disable the remote control for production use case.

Seems to me like having a dedicated health check queue is the right way but I'm not sure at all

Any one has any other direction not involving remote control to test health checks?

**Comment by bartoszhernas** [comment]
We use dedicated health checks queues with RabbitMQ eviction policies (queues are deleted automatically) successfully for some time now, and we are happy with the solution. Mostly because this check is indeed checking that worker processes the task and finishes. Since we've introduced it we had no more problems with stuck workers.

**Comment by innovia** [comment]
@bartoszhernas mind sharing some code for that? do you queue these via beat and than workers pick it up?

would love to see the code + the liveness probe section

**Comment by bartoszhernas** [comment]
Hi, the code is really easy:

In Kubernetes I specify queue name based on POD_NAME, and pass it to livecheck script:

```
        livenessProbe:
          initialDelaySeconds: 120
          periodSeconds: 70
          failureThreshold: 1
          exec:
            command:
            - bash 
            - "-c" 
            - |
              python celery_liveness_probe.py $LIVENESS_QUEUE_NAME
        env:
        - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name

        - name: LIVENESS_QUEUE_NAME
          value: queue-$(MY_POD_NAME)
```
(you need to use bash -c, because Kubernetes doesn't expand ENVs when trying to pass it as command directly)

then the celery_liveness_probe.py is just setting up Django to be able to use Celery and scheduling the task on POD's queue
```
# encoding: utf-8
from __future__ import absolute_import, unicode_literals

import os
import sys

if __name__ == "__main__":
    import django

    sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "ahoy.archive.settings")
    django.setup()
    from ahoy.archive.apps.eventbus.service import eventbus_service

    exit(0 if eventbus_service.health_check(sys.argv[1] if sys.argv and len(sys.argv) > 1 else None) else 1)
```

the health check function sends the task and waits for the results

```
    def health_check(self, queue_name: Optional[str] = None) -> bool:
        event = self.celery.send_task(
            AhoyEventBusTaskName.LIVENESS_PROBE,
            None,
            queue=queue_name or self.origin_queue,
            ignore_result=False,
            acks_late=True,
            retry=False,
            priority=255
        )

        try:
            is_success = event.get(timeout=10)
        except (celery.exceptions.TimeoutError, AttributeError):
            is_success = False

        return is_success

```

So basically: send a task, and if it returns results then the worker is healthy. If the worker got stuck (happened a lot of times) then the tasks never finishes, the Pod is restarted and everything gets back to normal.

The only caveat is that you need to deal with old queues, with RabbitMQ it's easy, we just set up expiry policy on queue
https://www.rabbitmq.com/ttl.html#queue-ttl

**Comment by innovia** [comment]
@bartoszhernas thank you for sharing the code! 

like you've said my queues are dynamic and we are using Redis - so we really need to find a way to deal with the queue names expiration on Redis

**Comment by bartoszhernas** [comment]
Yeah, we have similar issue with BullMQ with Redis. My idea is to write CronJob for Kubernetes that would clear the queues every some time.

**Comment by anithashok1910** [comment]
The goal of liveness for celery beat/scheduler is to check if the celery beat/scheduler is able to send the job to the message broker so that it can be picked up by the respective consumer. [Is it still working or in a hung state] The celery worker and celery scheduler/beat may or may not be running in the same pod or instance.

To handle such scenarios, we can create a method `update_scheduler_liveness` with decorator `@after_task_publish.connect ` which will be called every time when the scheduler successfully publishes the message/task to the message broker. 
The method `update_scheduler_liveness` will update the current timestamp to a file every time when the task is published successfully.
In Liveness probe,  we need to check the last updated timestamp of the file either using:
` stat --printf="%Y" celery_beat_schedule_liveness.stat ` command
or we can explicitly try to read the file (read mode) and extract the timestamp and compare if the the timestamp is recent or not based on the liveness probe criteria.
In this approach, the more minute liveness criteria you need, the more frequent a job must be triggered from the celery beat. So, for those cases, where the frequency between jobs is pretty huge, a custom/dedicated liveness heartbeat job can be scheduled every 2-5 mins and the consumer can just process it. `@after_task_publish.connect ` decorator provides multiple arguments that can be also used for filtering of liveness specific job that were triggered 

If we don't want to go for file based approach, then we can rely on Redis like data-source with instance specific redis key as well which needs to be implemented on the same lines.

**Comment by zffocussss** [comment]
inspect stats and inspect ping take too much time to response

**Comment by zffocussss** [comment]
> Beware that in some apps, running this command can take a few seconds using full CPU AND kubernetes defaults are to run it every 10 seconds.
> 
> It is thus much safer to have a high periodSeconds (ours is set to 300).

yes I met it

**Comment by GeorgiosNikolopoulos** [comment]
In our cluster, running `celery -A path.to.app inspect ping -d celery@$HOSTNAME` resulted in some pretty awful 100% CPU usage. Additionally, the command would sometimes completely fail, leading to restarted pods despite the worker actually working.

We instead tested and used `celery -b $CELERY__BROKER__URL -d celery@$HOSTNAME` and that resulted in only 30% CPU usage that also didn't fail randomly.  So, if you're also having a problem with -A, consider passing the broker url directly to the command via -b.

**Comment by beje2k15** [comment]
Has anybody tried to use [celery signals](https://docs.celeryproject.org/en/stable/userguide/signals.html#worker-signals) to solve this issue?

I was thinking of:

- readinessProbe
    Using [worker_ready](https://docs.celeryproject.org/en/stable/userguide/signals.html#worker-ready) to create and [worker_shutdown](https://docs.celeryproject.org/en/stable/userguide/signals.html#worker-shutdown) to delete a file (e.g. /tmp/worker_ready and let check k8s for file existence

- livenessProbe
  Using [heartbeat_sent](https://docs.celeryproject.org/en/stable/userguide/signals.html#heartbeat-sent) to write to a file (e.g. current timestamp) and probe it with something like `test $(($(date +%s) - $(stat -c %Y /tmp/worker_heartbeat))) -lt 10`

In my opinion this would incoporate celerys monitoring api pointed out by @thedrow and should not introduce excessive load like ping commands

**Comment by beje2k15** [comment]
I implemented my idea and so far it is working. However heartbeats are not send if the broker is unreachable (in my case redis), so livenessProbe will fail in this case causing the pods to be killed. I don't know if this actually a good thing or not.

code:
```python
# code for readiness and liveness probes
from celery.signals import worker_ready, worker_shutdown, heartbeat_sent
import time
from os import remove

HEARTBEAT_FILE = "/tmp/worker_heartbeat"
READINESS_FILE = "/tmp/worker_ready"

@heartbeat_sent.connect
def heartbeat(**kwargs):
    with open(HEARTBEAT_FILE, "w") as fp:
        fp.write(f"{time.time():.0f}\n")

@worker_ready.connect
def worker_ready(**kwargs):
    with open(READINESS_FILE, "w") as fp:
        fp.write(f"{time.time():.0f}\n")

@worker_shutdown.connect
def worker_shutdown(**kwargs):
    for fn in (HEARTBEAT_FILE, READINESS_FILE):
        remove(fn)
```

probes:

```yaml
livenessProbe:
  exec:
    command:
    - sh
    - -c
    - test $(($(date +%s) - $(stat -c %Y /tmp/worker_heartbeat))) -lt 10
  initialDelaySeconds: 10
  periodSeconds: 5
readinessProbe:
  exec:
    command:
    - test
    - -e /tmp/worker_ready
  initialDelaySeconds: 10
  periodSeconds: 5
```

Of course this code is now at project level. However I think this could be added to celery directly. Code could be added e.g. to `celery.utils.k8sprobes` (`celery.worker.k8sprobes`) so any project which need this probes just import the module to have the handlers installed. The module could also implement a main function which implements the probes, so k8s probes would just call `python -m celery.utils.k8sprobes livenessProbe` to do the check

**Comment by limedaniel** [comment]
For what it's worth, we used @beje2k15 's solution and it appears to be working very well.  We made a few changes to the Python portion though to simplify things (and leverage `pathlib`):

```python
from pathlib import Path

from celery.signals import heartbeat_sent, worker_ready, worker_shutdown


HEARTBEAT_FILE = Path("/tmp/worker_heartbeat")
READINESS_FILE = Path("/tmp/worker_ready")


@heartbeat_sent.connect
def heartbeat(**_):
    HEARTBEAT_FILE.touch()


@worker_ready.connect
def worker_ready(**_):
    READINESS_FILE.touch()


@worker_shutdown.connect
def worker_shutdown(**_):
    for f in (HEARTBEAT_FILE, READINESS_FILE):
        f.unlink()
```

**Comment by zffocussss** [comment]
can flower do the monitor tasks indirectly?

**Comment by taleinat** [comment]
> ```yaml
>     command:
>     - test
>     - -e /tmp/worker_ready
> ```

Note that the above Kubernetes configuration for the readiness probe from @beje2k15's comment has a bug, which causes the readiness probe to never fail. Here's one way to fix it:

```yaml
    command:
    - sh
    - -c
    - test -e /tmp/worker_ready
```

**Comment by joekohlsdorf** [comment]
The approach is good but it requires you to spam your broker with heartbeat garbage.
Here is an improved version which doesn't require heartbeats:

```python
from pathlib import Path

from celery import bootsteps
from celery.signals import worker_ready, worker_shutdown

HEARTBEAT_FILE = Path("/tmp/worker_heartbeat")
READINESS_FILE = Path("/tmp/worker_ready")

class LivenessProbe(bootsteps.StartStopStep):
    requires = {'celery.worker.components:Timer'}

    def __init__(self, worker, **kwargs):
        self.requests = []
        self.tref = None

    def start(self, worker):
        self.tref = worker.timer.call_repeatedly(
            1.0, self.update_heartbeat_file, (worker,), priority=10,
        )

    def stop(self, worker):
       HEARTBEAT_FILE.unlink(missing_ok=True)

    def update_heartbeat_file(self, worker):
       HEARTBEAT_FILE.touch()


@worker_ready.connect
def worker_ready(**_):
   READINESS_FILE.touch()


@worker_shutdown.connect
def worker_shutdown(**_):
   READINESS_FILE.unlink(missing_ok=True)


app = Celery("appname")
app.steps["worker"].add(LivenessProbe)
```

The liveness file will be created once the connection to the broker is successfully established so you could also use that for readiness.

**Comment by chicocvenancio** [comment]
> Since we are already checking the connection to the celery broker, does it still make sense to check the status of the worker?

In my experience, yes. Broker connection only guarantees the pod can connect to the broker, but its not uncomon for celery worker to be in a broken state even if that passes (stopped queue with rabbitmq is a scenario that comes to mind). We use a file do only do these checks every 5 minutes though, as they are a bit expensive.

**Comment by ishallbethat** [comment]
How about beat ? above healthcheck setting doesnt' seem to work with beat

**Comment by ahopkins** [comment]
You can do something similar where beat runs a "health" task and writes the timestamp to a file and you just check that.

**Comment by lachtanek** [comment]
as others have suggested, we've started using `celery -b $BROKER_URL -d celery@$HOSTNAME` for our health checks with a 60 second period and it seems not to raise the overall CPU usage too much (make sure not to use with `-A` though, as that'd load the whole app), I'd recommend to use that

**Comment by suciua** [comment]
> The approach is good but it requires you to spam your broker with heartbeat garbage. Here is an improved version which doesn't require heartbeats:
> 
> ```python
> from pathlib import Path
> 
> from celery import bootsteps
> from celery.signals import worker_ready, worker_shutdown
> 
> HEARTBEAT_FILE = Path("/dev/shm/worker_heartbeat")
> READINESS_FILE = Path("/dev/shm/worker_ready")
> 
> class LivenessProbe(bootsteps.StartStopStep):
>     requires = {'celery.worker.components:Timer'}
> 
>     def __init__(self, worker, **kwargs):
>         self.requests = []
>         self.tref = None
> 
>     def start(self, worker):
>         self.tref = worker.timer.call_repeatedly(
>             1.0, self.update_heartbeat_file, (worker,), priority=10,
>         )
> 
>     def stop(self, worker):
>        HEARTBEAT_FILE.unlink(missing_ok=True)
> 
>     def update_heartbeat_file(self, worker):
>        HEARTBEAT_FILE.touch()
> 
> 
> @worker_ready.connect
> def worker_ready(**_):
>    READINESS_FILE.touch()
> 
> 
> @worker_shutdown.connect
> def worker_shutdown(**_):
>    READINESS_FILE.unlink(missing_ok=True)
> 
> 
> app = Celery("appname")
> app.steps["worker"].add(LivenessProbe)
> ```
> 
> The liveness file will be created once the connection to the broker is successfully established so you could also use that for readiness.

@joekohlsdorf Thanks a lot for your solution to implement the probes! It might be worth mentioning that `/dev/shm` doesn't necessarily get removed upon pod restarts which has quite some implications on these probes (see https://github.com/kubernetes/kubernetes/issues/81001). Thus it would be better to use the path suggested by @limedaniel and store the files in `/tmp` which would lead to:
```python
HEARTBEAT_FILE = Path("/tmp/worker_heartbeat")
READINESS_FILE = Path("/tmp/worker_ready")
```

**Comment by nickjj** [comment]
I really like the signal approach + file checks and it's what I was planning to implement as well, then I Google'd around and was pleasantly surprised to find these solutions posted here.

Celery also has the option to write out a pid file when it starts. Is the `worker_ready` signal better because the pid file gets created before the worker is ready to do work? I casually skimmed the code base and didn't see much around the life cycle of when the pid file gets created vs some of the signals.

**Comment by GitRon** [comment]
Great approach! Finally something, that works and doesn't look hacky. 

Any thought on how to make the same for the `celery-beat`? I found a signal for start-up:

```python
@beat_init.connect
def beat_ready(**_):
    READINESS_FILE.touch()
```

But I just can't find a solution for the liveness check...

**Comment by nickjj** [comment]
Now that this is tagged for a milestone which hints it'll be in Celery at some point, how necessary is that timer liveness probe?

If the worker can write that file every second it's healthy, no doubt about it. But if a container is running 1 process and that process crashes then the container dies with it. In a Kubernetes context the pod will be removed and restarted. 

Where I'm going with this one is can we remove that entire `LivenessProbe` class and hook up a startup, liveness and readiness probe to only do this along with having the `worker_ready` and `worker_shutdown` signals as they are defined above.

```yml
    command:
    - sh
    - -c
    - test -e /tmp/worker_ready
```

My thought process there is the file either always exists (the worker is working) or the process is dead in which case the file won't be there because the container isn't up.

**Comment by ibachar-es** [comment]
It seems the heartbeat using the file based approach is not working when using the gevent pool when the worker is blocking. Does anyone have a solution that works for gevent as well?

**Comment by auvipy** [comment]
have anyone tried keda?

**Comment by sondrelg** [comment]
> have anyone tried keda?

We're about to, for scaling workers by queue size. We just need to switch brokers to rabbitmq first. 

How does KEDA relate to readiness/liveness probes?

**Comment by benedikt-bartscher** [comment]
@some1ataplace please do us all a favour and don't just post random chatgpt crap on github

**Comment by SnoozeFreddo** [comment]
I also can't configure beat. There is an init without cleanup option (on_shutdown) and no liveness check Anyone figured this out?

**Comment by grapo** [comment]
We tried a lot of ideas proposed here like @beje2k15 solution with heartbeat signals. Sadly due to bugs like https://github.com/celery/celery/discussions/7276 worker can respond to heartbeats (so `heartbeat_sent` signal is catched) but still don't receive real tasks from broker.

Only fully reliable solution for us is to send own heartbeat task with celery beat, handle it with celery worker and use `task_success` signal to  touch liveness file. **This solution is only good if you have only one worker!** If there are more workers they may not get tasks from broker and fail liveness probe

```python
from pathlib import Path

from celery.signals import (
    after_task_publish,
    beat_init,
    task_success,
    worker_ready,
    worker_shutdown,
)

from .tasks import celery_heartbeat

HEARTBEAT_FILE = Path("/tmp/celery_live")
READINESS_FILE = Path("/tmp/celery_ready")

######
# celery worker rediness and liveness checks

@worker_ready.connect
def worker_ready(**_):
    READINESS_FILE.touch()


@worker_shutdown.connect
def worker_shutdown(**_):
    for f in (HEARTBEAT_FILE, READINESS_FILE):
        f.unlink(missing_ok=True)


@task_success.connect(sender=celery_heartbeat)
def heartbeat(**_):
    HEARTBEAT_FILE.touch()

######
# celery beat rediness and liveness checks

@beat_init.connect
def beat_ready(**_):
    READINESS_FILE.touch()


@after_task_publish.connect(sender="healthcheck.tasks.celery_heartbeat")
def task_published(**_):
    HEARTBEAT_FILE.touch()

```

This is our code for liveness probe. We check if  `/tmp/celery_live` was touched in the last minute.

```yaml
          livenessProbe:
            initialDelaySeconds: 90
            periodSeconds: 30
            failureThreshold: 3
            timeoutSeconds: 3
            exec:
              command:
                - /bin/sh
                - c
                - find  /tmp/celery_live -mmin -1 | grep .
```

**Comment by GitRon** [comment]
@grapo Thx for the detailed answer but I think that your case of only having one worker won't do for most of the crowd. You have celery to be able to scale, right? Nevertheless, thx for sharing!

**Comment by sbasu-mdsol** [comment]
What @GitRon said. I am working on a scenario where there will be at least 4-6 workers and I need to have a readiness probe that indicates that all of them are ready to process tasks.

**Comment by joekohlsdorf** [comment]
Testing if the broker is alive from the worker makes no sense to me. It will lead to all workers restarting unnecessarily if the broker has a temporary issue which is unrelated to Celery workers. Workers can recover from broker failure without restart.
The Celery worker liveness check should only monitor if the worker is ready to accept tasks which is achieved by the code I shared in https://github.com/celery/celery/issues/4079#issuecomment-1128954283

If you really want full roundtrip monitoring you can use the `celery inspect ping` command I shared previously but it uses more CPU. This sends a message to the individual pidbox of the worker which goes through the broker. You can probably implement what the ping command does with a timer in the worker itself to optimize this.

All of these checks still don't mean that you are able to process real tasks but there is really no good way to check this on the worker. You can count processed tasks but that doesn't mean anything because your system might only process one task per month. I recommend to also monitor queue size and ack rate on your broker.

Timers can be used to do all kinds of cool things, I have code which emits currently active tasks and general worker statistics as statsd metics. This is way better than Flower or inspect.

**Comment by rbehal** [comment]
This worked best for me

```
          livenessProbe:
            exec:
              command:
                - sh
                - -c
                - celery inspect ping -d celery@$(hostname) | grep -q OK
            initialDelaySeconds: 30
            periodSeconds: 15
            failureThreshold: 3
          readinessProbe:
            exec:
              command:
                - sh
                - -c
                - celery inspect ping -d celery@$(hostname) | grep -q OK
            initialDelaySeconds: 60
            periodSeconds: 15
            failureThreshold: 3
```

It relies on your base image being able to do `$(hostname)` though

**Comment by rblack** [comment]
The current behavior is that even a brief disconnect to the broker causes this custom bootstep [#4079 (comment)](https://github.com/celery/celery/issues/4079#issuecomment-1128954283) to stop working. As a result, any disconnect will necessarily lead to a pod restart. For more information, see issue #3944, which also includes steps to reproduce the problem. Testing done using latest currently available celery 5.4.0, but judging the #3944 is from 2017, it's not only the latest Celery affected.

Unless I am missing something, under these circumstances, it is better to set BROKER_CONNECTION_MAX_RETRIES=0 to restart the pod immediately on disconnect, rather than waiting for the livenessProbe to eventually kill it anyway.

**Comment by famagusta** [comment]
@rblack Did you try switching to consumer bootstep instead of worker bootstep to see if it resolves the issue (as suggested in #3944 )? We are facing the same issue of liveliness failing on RMQ disconnects and were considering switching to Consumer bootstep to resolve it

**Comment by rblack** [comment]
@famagusta Yes, I tried switching to consumer level and it resolves the issue. I didn't dive deep into celery architecture. However, as far as I understand, where even a working **worker** bootstep doesn't guarantee that the worker is healthy and able to process tasks, the **consumer** bootstep provides even less guarantees. So as I mentioned earlier, worker bootstep and trade-off using [k8s restart](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy) failed pod instead of celery's built-in reconnection mechanism, in my opinion is more reliable than switching to consumer level.

**Comment by WillNilges** [comment]
Has anyone gotten [this solution](https://github.com/celery/celery/issues/4079#issuecomment-1270085680) with signals working with MyPy yet? I'm struggling to wrap my head around how exactly it works, doubly so with what type the `worker`s are supposed to be. I got it to work, but I just had to disable mypy for that file.

**Comment by BITSOLVER** [comment]
I am having this issue with celery 5.4.0 amqp 5.2.0 kombu 5.4.2 redis 5.1.1 (redis server 7.4).

Worker/beat running:
 -------------- celery@AF40CBB.online-server.cloud v5.4.0 (opalescent)
--- ***** ----- 
-- ******* ---- Linux-4.18.0-553.22.1.el8_10.x86_64-x86_64-with-glibc2.28 2024-10-19 12:54:59
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         DAVO:0x7f88289b0560
- ** ---------- .> transport:   redis://localhost:6379//
- ** ---------- .> results:     redis://localhost:6379/
- *** --- * --- .> concurrency: 24 (prefork)
-- ******* ---- .> task events: ON
--- ***** ----- 
 -------------- [queues]
                .> steven_DAVO_Manager exchange=steven_DAVO_Manager(direct) key=steven_DAVO_Manager

Running inpect:
(venv3-12) [steven@AF40CBB src]$ celery -A DAVO inspect ping  -t 5
[2024-10-19 13:47:07,218] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:40 BASE_DIR: /home/steven/PycharmProjects/davo/Django/src
[2024-10-19 13:47:07,218] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:45 Test output folder: /home/steven/PycharmProjects/davo/TestOutput
[2024-10-19 13:47:07,218] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:48 Default unit root: /home/steven/PycharmProjects/davo/Django/src/DAVO/tests
[2024-10-19 13:47:07,218] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:55 Load units from: /home/steven/PycharmProjects/davo/Django/src/DAVO/tests
[2024-10-19 13:47:07,250] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:269 Celery queue: steven_DAVO_Manager
[2024-10-19 13:47:07,250] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:301 CORS_ALLOWED_ORIGINS: ['https://storage.googleapis.com', 'http://api-davo.localhost:8070', 'http://dev-davo.localhost:8085', 'http://localhost:8070', 'http://localhost:8085', 'http://localhost:8285']
[2024-10-19 13:47:07,250] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:307 CORS_ALLOW_ALL_ORIGINS: True
[2024-10-19 13:47:07,250] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:309 CORS_ALLOW_CREDENTIALS: True
[2024-10-19 13:47:07,250] INFO: /home/steven/PycharmProjects/davo/Django/src/DAVO/settings.py:315 CORS_ALLOW_HEADERS: ['accept', 'authorization', 'content-type', 'user-agent', 'x-csrftoken', 'x-requested-with', 'xs-location']
Error: No nodes replied within time constraint

**Comment by RRRajput** [comment]
@rblack, can you please specify what needs to change in this [implementation](https://github.com/celery/celery/issues/4079#issuecomment-1270085680) in order to use consumer bootstep?

**Comment by rblack** [comment]
@RRRajput, remove timer requirement  due to #3955 and add step as in [docs](https://docs.celeryq.dev/en/latest/userguide/extending.html#installing-bootsteps). I'm not sure when it will fail without the Timer requirement but it works now. 

```diff
class LivenessProbe(bootsteps.StartStopStep):
-   requires = {'celery.worker.components:Timer'}

    def __init__(self, worker, **kwargs):
        self.requests = []
        self.tref = None

...

app = Celery("appname")
- app.steps["worker"].add(LivenessProbe)
+ app.steps["consumer"].add(LivenessProbe)
```

**Comment by zhanqian-zhang** [comment]
https://github.com/apache/airflow/blob/d2449821512c81b828c5022d224b273f172384f6/providers/celery/src/airflow/providers/celery/cli/celery_command.py#L273
Referring to the Airflow worker implementation, maybe we can run a another process to expose an endpoint for health checks.

**Comment by wosc** [comment]
I also guess the concrete solution depends on what exactly you want to achieve with the probe. My main usecase so far is "keep the old code running if we deploy a broken update" (and not so much "detect if a running worker does not process tasks anymore"), which I solved by counting worker processes (similar to [this comment](https://github.com/celery/celery/issues/4079#issuecomment-1041188705)). I've packaged that code as https://pypi.org/project/celery_worker_healthcheck/

**Comment by gertzakis** [comment]
I was using files similar to [this comment](https://github.com/celery/celery/issues/4079#issuecomment-1270085680) and it was working like a charm until Celery `5.6.1`. When we upgraded celery to `5.6.1` the heartbeat worker files are not created any more, even though we didn't change anything. Any ideas? Or anyone else experiencing the same?

**Comment by melbic** [comment]
@gertzakis I am interested in your solution, but unfortunately your link is broken

**Reply by gertzakis** [reply]
Updated it with the proper link! Thanks for the heads up
